### PR TITLE
add opaque type support to hive type serde

### PR DESCRIPTION
**Pull request by kunigami**
Summary:
My understanding of opaque types in Velox is that Velox doesn't know about the underlying type of it, and treats them as a `shared_ptr<void>`. For serializing data across processes, we need to somewhat break that assumption, because when we need to know how to deserialize this opaque data.

One option is to have the underlying type as part of the serialized type signature, the other is to store this information with the serialized data itself. I'm adopting the first option here.

We also need to introduce a layer of abstraction for opaque type index, by allowing aliasing opaque types. 

The reason we can't use opaque type index is the assumption that they're not stable across processes. So if you serialize a opaque type as string in process A and then deserialize in process B, even if running the same revision there's no guarantee the type ID is the same.

With this change, callers are required to register an alias for opaque types before serializing/deserializing it via `HiveTypeSerializer` and `HiveTypeParser`. I put this registry in `Type.h` but if we want to keep this specific to `HiveTypeSerializer/HiveTypeParser` we could move it elsewhere.

Differential Revision: D64358220

### CODE REVIEW

File: velox/type/Type.cpp [hunk 1]
```diff
@@ -1206,4 +1218,22 @@ TypePtr getType(
   return getCustomType(name);
 }
 
+std::type_index getTypeIdForOpaqueTypeAlias(const std::string& name) {
+  auto it = getTypeIndexByOpaqueAlias().find(name);
+  VELOX_CHECK(
+      it != getTypeIndexByOpaqueAlias().end(),
+      "Could not find type {}. Did you call registerOpaqueType?",
```
Comment (pedroerp) [inline velox/type/Type.cpp:1225]: nit: please surround the type name by quotes `'{}'`

File: velox/type/Type.h [hunk 2]
```diff
@@ -2008,6 +2008,22 @@ bool registerCustomType(
     const std::string& name,
     std::unique_ptr<const CustomTypeFactories> factories);
 
+std::unordered_map<std::string, std::type_index>& getTypeIndexByOpaqueAlias();
```
Comment (pedroerp) [inline velox/type/Type.h:2011]: can you add a bit more documentation here about what this does, when it should be used, and some of the restrictions/considerations you added to the PR summary?

File: velox/type/fbhive/HiveTypeParser.cpp [hunk 3]
```diff
@@ -87,7 +88,16 @@ std::shared_ptr<const Type> HiveTypeParser::parse(const std::string& ser) {
 Result HiveTypeParser::parseType() {
   Token nt = nextToken();
   VELOX_CHECK(!nt.isEOS(), "Unexpected end of stream parsing type!!!");
-  if (nt.isValidType() && nt.isPrimitiveType()) {
+
+  if (!nt.isValidType()) {
+    VELOX_FAIL(fmt::format(
```
Comment (pedroerp) [inline velox/type/fbhive/HiveTypeParser.cpp:93]: I know you're only copying it, but you can omit the fmt::format as Velox exception macros can do the formatting automatically for you.

File: velox/type/fbhive/HiveTypeParser.cpp [hunk 4]
```diff
@@ -118,7 +128,16 @@ Result HiveTypeParser::parseType() {
       eatToken(TokenType::RightRoundBracket);
     }
     return Result{scalarType};
-  } else if (nt.isValidType()) {
+  } else if (nt.metadata->tokenString[0] == "opaque") {
```
Comment (pedroerp) [inline velox/type/fbhive/HiveTypeParser.cpp:131]: should we have a `nt.isOpaqueType()`?

File: velox/type/fbhive/tests/HiveTypeParserTests.cpp [hunk 5]
```diff
@@ -176,4 +176,13 @@ TEST(FbHive, parseSpecialChar) {
   ASSERT_EQ(t->toString(), "ROW<\"a$_#\":INTEGER>");
 }
 
+struct Foo {};
+TEST(FbHive, parseOpaque) {
+  // Use a custom name to highlight this is just an alias.
+  registerOpaqueType<Foo>("bar");
+  HiveTypeParser parser;
+  auto t = parser.parse("opaque<bar>");
+  ASSERT_EQ(t->toString(), "OPAQUE<facebook::velox::type::fbhive::Foo>");
```
Comment (pedroerp) [inline velox/type/fbhive/tests/HiveTypeParserTests.cpp:185]: very cool :)

File: velox/type/fbhive/tests/HiveTypeParserTests.cpp [hunk 6]
```diff
@@ -176,4 +176,13 @@ TEST(FbHive, parseSpecialChar) {
...
+}
```
Comment (pedroerp) [inline velox/type/fbhive/tests/HiveTypeParserTests.cpp:186]: should we also test an invalid opaque type deserialization?

File: velox/type/fbhive/tests/HiveTypeSerializerTests.cpp [hunk 7]
```diff
@@ -33,9 +33,19 @@ TEST(HiveTypeSerializer, primitive) {
   EXPECT_EQ(result, "bigint");
 }
 
+struct Foo {};
 TEST(HiveTypeSerializer, opaque) {
-  std::shared_ptr<const velox::Type> type = velox::OPAQUE<bool>();
+  // Use a custom name to highlight this is just an alias.
+  registerOpaqueType<Foo>("bar");
+
+  std::shared_ptr<const velox::Type> type = velox::OPAQUE<Foo>();
```
Comment (pedroerp) [inline velox/type/fbhive/tests/HiveTypeSerializerTests.cpp:41]: you can omit the `velox::` namespace prefix throughout. You can also use the TypePtr alias or just auto

File: velox/type/fbhive/tests/HiveTypeSerializerTests.cpp [hunk 8]
```diff
@@ -33,9 +33,19 @@ TEST(HiveTypeSerializer, primitive) {
...
+  auto result = HiveTypeSerializer::serialize(type);
+  EXPECT_EQ(result, "opaque<bar>");
+}
+
+TEST(HiveTypeSerializer, unsupported) {
```
Comment (pedroerp) [inline velox/type/fbhive/tests/HiveTypeSerializerTests.cpp:46]: can we also test some unregistered type inside the opaque?

File: velox/vector/fuzzer/VectorFuzzer.cpp [hunk 9]
```diff
@@ -558,6 +558,8 @@ VectorPtr VectorFuzzer::fuzzFlat(const TypePtr& type, vector_size_t size) {
     }
 
     return fuzzRow(std::move(childrenVectors), rowType.names(), size);
+  } else if (type->isOpaque()) {
```
Comment (pedroerp) [inline velox/vector/fuzzer/VectorFuzzer.cpp:561]: Cc: @kagamiori @bikramSingh91 I remember some discussion about adding opaque support in Fuzzer; here it is :)

@kunigami could you just split the fuzzer support stuff to a different PR?

File: velox/vector/fuzzer/VectorFuzzer.cpp [hunk 10]
```diff
@@ -629,6 +631,27 @@ VectorPtr VectorFuzzer::fuzzComplex(const TypePtr& type, vector_size_t size) {
   return nullptr; // no-op.
 }
 
+VectorPtr VectorFuzzer::fuzzFlatOpaque(
+    const TypePtr& type,
+    vector_size_t size) {
+  auto vector = BaseVector::create(type, size, pool_);
+  using TFlat = typename KindToFlatVector<TypeKind::OPAQUE>::type;
```
Comment (pedroerp) [inline velox/vector/fuzzer/VectorFuzzer.cpp:638]: nit: maybe `TFlatOpaque` ?

File: velox/vector/fuzzer/VectorFuzzer.cpp (hunk 9 above)
Comment (kunigami) [reply velox/vector/fuzzer/VectorFuzzer.cpp:561]: I actually had a PR for adding it to fuzzer but I missed some codepaths that this test triggered. I'll move it to
https://github.com/facebookincubator/velox/pull/11189

Comment (kunigami) [reply velox/vector/fuzzer/VectorFuzzer.cpp:561]: nvm, I just messed the export of the diff stack.

File: velox/type/Type.h [hunk 11]
```diff
@@ -2009,6 +2009,34 @@ bool registerCustomType(
...
+
+// Reverse of getTypeIndexByOpaqueAlias() when we need to look up the opaque
+// alias by its type index.
+std::unordered_map<std::type_index, std::string>& getOpaqueAliasByTypeIndex();
+
+std::type_index getTypeIdForOpaqueTypeAlias(const std::string& name);
+
+std::string getOpaqueAliasForTypeId(std::type_index typeIndex);
+
+// OpaqueType represents a type that is transparent to the Velox type system.
```
Comment (pedroerp) [inline velox/type/Type.h:2024]: nit: I think the point is that opaque types are NOT transparent to the type system. :) (they are "opaque" to the type system) :)

File: velox/type/Type.h [hunk 12]
```diff
@@ -2009,6 +2009,34 @@ bool registerCustomType(
...
+// To identify the underlying type we use std::type_index which is stable
+// within the same process. However, it is not necessarily stable across
+// processes.
+//
+// So if we were to serialize an opaque type using its std::type_index, we
+// might not be able to deserialize it in another process. To solve this
+// problem, we require that both the serializing and deserializing processes
+// register the opaque type using registerOpaqueType() with the same alias.
```
Comment (pedroerp) [inline velox/type/Type.h:2032]: thanks for the comment. We usually use "///" for documentation header comments such as this.

File: velox/type/fbhive/HiveTypeParser.cpp [hunk 13]
```diff
@@ -140,11 +159,6 @@ Result HiveTypeParser::parseType() {
       default:
```
Comment (pedroerp) [inline velox/type/fbhive/HiveTypeParser.cpp:159]: could we use `toString(nt.typeKind())` instead?

File: velox/type/fbhive/tests/HiveTypeParserTests.cpp (hunk 6 above)
Comment (pedroerp) [reply velox/type/fbhive/tests/HiveTypeParserTests.cpp:186]: is it on purpose that we still don't have deserialization tests?

**Review by pedroerp — APPROVED** [review]
Thanks @kunigami , looks pretty good!

**Comment by pedroerp** [comment]
(please rebase on main first) @kunigami

File: velox/type/fbhive/tests/HiveTypeParserTests.cpp (hunk 6 above)
Comment (kunigami) [reply velox/type/fbhive/tests/HiveTypeParserTests.cpp:186]: not sure I understand, what deserialization tests?

**Comment by kunigami** [comment]
rebased
